source/server/http/admin.cc

```diff
@@ -514,6 +514,50 @@ Http::Code AdminImpl::handlerCpuProfiler(absl::string_view url, Http::HeaderMap&
   return Http::Code::OK;
 }
 
+Http::Code AdminImpl::handlerHeapProfiler(absl::string_view url, Http::HeaderMap&,
+                                          Buffer::Instance& response, AdminStream&) {
+  if (!Profiler::Heap::profilerEnabled()) {
+    response.add("The current build does not support heap profiler");
+    return Http::Code::NotImplemented;
+  }
+
+  Http::Utility::QueryParams query_params = Http::Utility::parseQueryString(url);
+  if (query_params.size() != 1 || query_params.begin()->first != "enable" ||
+      (query_params.begin()->second != "y" && query_params.begin()->second != "n")) {
+    response.add("?enable=<y|n>\n");
+    return Http::Code::BadRequest;
+  }
+
+  Http::Code res = Http::Code::OK;
+  bool enable = query_params.begin()->second == "y";
+  if (enable) {
+    if (Profiler::Heap::isProfilerStarted()) {
+      response.add("Fail to start heap profiler: already started");
+      res = Http::Code::BadRequest;
+    } else if (!Profiler::Heap::startProfiler(profile_path_)) {
+      NOT_IMPLEMENTED_GCOVR_EXCL_LINE
+      response.add("Fail to start the heap profiler");
+      res = Http::Code::InternalServerError;
+    } else {
+      response.add("Starting heap profiler");
+      res = Http::Code::OK;
+    }
+  } else {
+    // !enable
+    if (!Profiler::Heap::isProfilerStarted()) {
+      response.add("Fail to stop heap profiler: not started");
+      res = Http::Code::BadRequest;
+    } else {
+      Profiler::Heap::stopProfiler();
+      response.add(fmt::format("Heap profiler stopped. Please checkout {}. See "
```

Comment:
language nit:

Heap profiler stopped and data written to {}. See http://goog-perftools.sourceforge.net/doc/heap_profiler.html for details.